### ISSUE TITLE
do not replace textarea value on paste event

### ISSUE THREAD

**Issue by Bogdan1001** [phoenixframework/phoenix_live_view#1117]
### Environment

* Elixir version (Elixir 1.10.4):
* Phoenix version (1.5.1):
* Phoenix LiveView version 0.14.4):
* NodeJS version (v12.18.2):
* NPM version (6.14.5):
* Operating system: Mac OS Catalina
* Browsers you attempted to reproduce this bug on (the more the merrier):  Brave/Chrome
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: Yes

### Actual behavior

Hello,

When I enter text to codemirror manually - it correctly updates the textarea.

But when I try to paste(cmd + v) some value to codemirror - I do not get any update . 

Here is how my Hook looks like:

```
Hooks.code_mirror = {
  mounted () {
const editor = CodeMirror.fromTextArea(document.getElementById('cm_textarea'),
    {
      mode: "text/html",
      lineNumbers: true,
      theme: "material"
    });

    editor.on('change', (editor) => {
      const text = editor.getValue()
      console.log(text)

      editor.save();
      // this.pushEvent("change", {csv: text})
    })
  }
}
```

At the same time, in console.log(text), I can see the pasted value in the console. 

Also I get  value correctly on change by using:  `this.pushEvent("change", {csv: text})`

Here is my form:
```
    <%= form_for @changeset, "#",
            phx_change: "change" %>

    <textarea id="cm_textarea" name="textarea_value" phx-hook="code_mirror">
    <%= @textarea_value %>
    </textarea>
```
### Expected behavior

Get the pasted data on change, from the `editor.save()` in my LV form module.

Related issue: https://github.com/codemirror/CodeMirror/issues/6378

**Comment by josevalim** [comment]
Please try adding phx-update=ignore to the text area so it ignores the server values. If that doesn’t work, then please provide a sample live view (module + template) that reproduces the error, as well as the JS files. Thanks!

**Comment by Bogdan1001** [comment]
Yes, phx-update="ignore" already was there. And this option does not help solve the issue.

```
use EmailMixerWeb, :live_view

  @impl true
  def mount(_params, _session, socket) do

    {:ok, socket}
  end

  @impl true
  def render(assigns) do
    ~L"""
    <div id="code_hook" phx-hook="code_mirror" phx-update="ignore" class="shadow-xl border border-gray-900">
    <%= f = form_for EmailMixer.List.changeset(%{}), "#",
            phx_change: "change"%>
            
      <%= textarea f, :code_mirror, placeholder: "Copy", id: "cm_textarea" %>
    <div>

    """
  end

  @impl true
  def handle_event("change", params, socket) do
    IO.inspect params
    {:noreply, socket}
  end
```
```

<!DOCTYPE html>
<html lang="en"
      class="h-full">

<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible"
        content="IE=edge" />
  <meta name="viewport"
        content="width=device-width, initial-scale=1.0" /> <%= csrf_meta_tag() %> <%= live_title_tag assigns[:page_title] ||
    "ModalExample", suffix: "" %>
  <link phx-track-static
        rel="stylesheet"
        href="<%= Routes.static_path(@conn, "/css/app.css") %>" />
  <script defer
          phx-track-static
          type="text/javascript"
          src="<%= Routes.static_path(@conn, "/js/app.js") %>" />
  </script>
      <script src="https://cdn.ckeditor.com/ckeditor5/21.0.0/classic/ckeditor.js"></script>

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.56.0/codemirror.min.css" integrity="sha512-/BlxZbYLtYGExexketXsTi47eHp+r2kTeq2OHecQPZlfbc7WFXVrwbVW9HOYjI6c9Ti+P60ASmVLxittZ0EBGw==" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.56.0/codemirror.min.js" integrity="sha512-H5RyE1jRbGn8d0dsSstZi0SSs23X6Hkkf5hLneEIFyCfP/cBwJ/jFJESBuFBiUyzbGZGa06tnrvw1SboRKdwcA==" crossorigin="anonymous"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.56.0/theme/dracula.min.css" integrity="sha512-gFMl3u9d0xt3WR8ZeW05MWm3yZ+ZfgsBVXLSOiFz2xeVrZ8Neg0+V1kkRIo9LikyA/T9HuS91kDfc2XWse0K0A==" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.56.0/theme/material.min.css" integrity="sha512-t93jzyjNejvKJ0LbK+VuLVp89Y8olz3sMmGPhKXYsTjSLuu3a3dVaZljCK6g04ShmsANEM8xe/OB+IaBE2LRBA==" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.56.0/addon/display/placeholder.min.js" integrity="sha512-lEgnL1kJ21Hrgt7wckcvCHqyrFn0LwqSXBL0us8edspdonYvPsMiykeLgQOUdUwlF49dNhttYpiZ0g1eMG/nMw==" crossorigin="anonymous"></script>

<script src="https://cdn.jsdelivr.net/gh/alpinejs/alpine@v2.3.5/dist/alpine.min.js" defer></script>
</head>

<body>
</body>
```

**Comment by Bogdan1001** [comment]
and yes here is my app.js file. Do I need to provide something else?

```
import "../css/app.scss"

import "phoenix_html"
import {
  Socket
} from "phoenix"
import NProgress from "nprogress"
import {
  LiveSocket
} from "phoenix_live_view"
import "alpinejs"
import { PhoenixLiveViewDropzone } from "phoenix_live_view_drop_zone";

const Hooks = {}

Hooks.PhoenixLiveViewDropzone = new PhoenixLiveViewDropzone();

Hooks.filepond = {

  mounted() {
    FilePond.parse(document.body);
  }
}

Hooks.code_mirror = {

  mounted () {
const editor = CodeMirror.fromTextArea(document.getElementById('cm_textarea'),
    {
      mode: "text/html",
      lineNumbers: true,
      theme: "material"
    });

    
    editor.on('change', (editor) => {
      // editor.save();
      const text = editor.getValue()
      this.pushEvent("change_textarea", text)
    })
  }
}


Hooks.text_editor = {
  mounted() {
    console.log("test")
    ClassicEditor
    .create( document.querySelector( '#editor' ), {
      removePlugins: [ 'Heading', 'Insertmedia', 'Link' ],
      toolbar: [ 'bold', 'italic', 'bulletedList', 'numberedList' ]
    } )
    .then( editor => {
            console.log( editor );
    } )
    .catch( error => {
            console.error( error );
    } );
  }
}

Hooks.Modal = {
  mounted() {
    window.modalHook = this
  },
  destroyed() {
    window.modalHook = null
  },
  modalClosing(leaveDuration) {
    // Inform modal component when leave transition completes.
    setTimeout(() => {
      var selector = '#' + this.el.id
      if (document.querySelector(selector)) {
        this.pushEventTo(selector, 'modal-closed', {})
      }
    }, leaveDuration);
  }
}

Hooks.ConnectionStatus = {
  mounted() {
    window.connected = true
  },
  disconnected() {
    window.connected = false
  },
  reconnected() {
    window.connected = true
  }
}

let csrfToken = document.querySelector("meta[name='csrf-token']").getAttribute("content");
let liveSocket = new LiveSocket("/live", Socket, {hooks: Hooks, params: {_csrf_token: csrfToken}});


// Show progress bar on live navigation and form submits
window.addEventListener("phx:page-loading-start", info => NProgress.start())
window.addEventListener("phx:page-loading-stop", info => NProgress.done())

// connect if there are any LiveViews on the page
liveSocket.connect()

// liveSocket.enableDebug()
// liveSocket.enableLatencySim(1000)
window.liveSocket = liveSocket
```

**Comment by sfusato** [comment]
Played a bit with this. The `textarea` field is hidden by the JS library (CodeMirror): `style="display: none;"`. When you are editing, you are not actually editing the textarea value, but rather rely on CodeMirror to push any event to the `textarea` DOM element as well. Remember that LiveView only listens to that DOM element.

Haven't looked into CodeMirror's code to see how they update the `textarea`, but backspace (erasing) doesn't 'work' either.

Here's a quick hack I came up with to fix your issue:
```javascript
editor.on('change', (instance, changeObj) => {
  editor.save()
  if (changeObj.origin === "paste" || changeObj.origin === "+delete") {
    document.getElementById('cm_textarea').dispatchEvent(new Event('change', { 'bubbles': true }));
  }
})
```
Basically, check the change origin, and if it's from a `paste` or `delete` event, manually push a `change` event to your `textarea` element which LiveView will pick up.

**Comment by josevalim** [comment]
Thanks for the @sfusato for the detailed reply, fantastic work as usual! Closing per the above!
